arangod/Replication/Syncer.cpp

```diff
@@ -130,20 +138,43 @@ arangodb::Result applyCollectionDumpMarkerInternal(
     options.indexOperationMode = arangodb::Index::OperationMode::internal;
 
     try {
-      // try insert first
-      OperationResult opRes = trx.insert(coll->name(), slice, options);
+      OperationResult opRes;
+      bool useReplace = false;
+      VPackSlice keySlice = arangodb::transaction::helpers::extractKeyFromDocument(slice);
+
+      // if we are about to process a single document marker (outside of a multi-document
+      // transaction), we first check if the target document exists. if yes, we don't
+      // try an insert (which would fail anyway) but carry on with a replace.
+      if (trx.isSingleOperationTransaction() && keySlice.isString()) {
+        arangodb::LocalDocumentId const oldDocumentId =
+            coll->getPhysical()->lookupKey(&trx, keySlice);
+        if (oldDocumentId.isSet()) {
```

Comment:
why are you not just using the `options.overwrite` optimization you wrote ?

Reply:
its the exact same logic

Reply:
The problem is that `overwrite` only handles a unique constraint violation in the primary index.
However, if there is a unique constraint violation in a _secondary_ index, then the overwrite will still fail and not return us any valuable information...
That's why I moved away from initially using the overwrite functionality here.